### PR TITLE
Add ARM support

### PR DESCRIPTION
**Pull request by antroseco**
64-bit support is almost ready. 32-bit ARM is a bit harder to work on as we can't really run our tests yet.

There's still a bit of work to do on the runtimes.

### CODE REVIEW

File: std/aarch64_linux/syscalls.c3 [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    return _syscall_3(fd, iov_addr, iovcnt, /* writev */ 66);
+}
+
+function sys_mmap : (addr: isize, length : isize, prot: u32, flags: u32, fd: i32, offset: isize) -> iptr = {
+    return _syscall_6(addr, length, as_arithmetic(prot), as_arithmetic(flags), fd, offset, /* mmap */ 222);
+}
+
+function sys_munmap : (addr: isize, length : isize) -> void = {
+    _syscall_2(addr, length, /* munmap */ 215);
+}
```
Comment (DaveDuck321) [inline std/aarch64_linux/syscalls.c3:32]: Can we fix #99 with this PR?

File: std/arm_linux/runtime.S [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+.global _syscall_1
+_syscall_1:
+    # Syscall number is passed via r7, but r7 must be preserved, so save it on
+    # the stack.
+    push    {r7}
+    mov     r7, r1
+    swi     0
+    pop     {r7}
+    bx      lr
```
Comment (DaveDuck321) [inline std/arm_linux/runtime.S:53]: Are r4-r7 caller saved by arm/EABI? If not they should also be saved here

File: std/sys/mman.c3 [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+function SYS_PROT_READ  : () -> u32 = { return 0x1; } // Page can be read.
+function SYS_PROT_WRITE : () -> u32 = { return 0x2; } // Page can be written.
+function SYS_PROT_EXEC  : () -> u32 = { return 0x4; } // Page can be executed.
+function SYS_PROT_NONE  : () -> u32 = { return 0x0; } // Page can not be accessed.
+
+function SYS_MAP_SHARED          : () -> u32 = { return 0x1; } // Share changes.
+function SYS_MAP_PRIVATE         : () -> u32 = { return 0x2; } // Changes are private.
+function SYS_MAP_SHARED_VALIDATE : () -> u32 = { return 0x3; } // Share changes and validate extension flags.
+
+function SYS_MAP_ANONYMOUS : () -> u32 = { return 0x20; } // Don't use a file.
```
Comment (DaveDuck321) [inline std/sys/mman.c3:15]: This is much cleaner, nice

File: tests/arrays/alignment_aarch64_linux.c3 [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+    // Arrays use the alignment of their elements.
+    let small : int[3];
+
+    // Irrespective of the total array size.
+    let large : int[4];
+
+    return 0;
+}
+
+/// @FOR aarch64_linux
```
Comment (DaveDuck321) [inline tests/arrays/alignment_aarch64_linux.c3:11]: Ooo, smart. Depending on how many of these we want, a target directory might be nice. This is good for now tho.

File: tests/misc/fork.c3 [hunk 5]
```diff
@@ -14,6 +14,7 @@ function main : () -> int = {
     return 0;
 }
 
+/// @FOR x86_64_linux
```
Comment (DaveDuck321) [inline tests/misc/fork.c3:17]: Is this just temporary until we've implemented a `clone3` based `fork` for arm?

**Review by DaveDuck321 — APPROVED** [review]
Very cool

File: std/arm_linux/runtime.S (hunk 2 above)
Comment (antroseco) [reply std/arm_linux/runtime.S:53]: r4-r11 and sp are callee-saved, but we are only touching r7, right?

File: tests/misc/fork.c3 (hunk 5 above)
Comment (antroseco) [reply tests/misc/fork.c3:17]: Yes, although that's not really a priority

File: tests/arrays/alignment_aarch64_linux.c3 (hunk 4 above)
Comment (antroseco) [reply tests/arrays/alignment_aarch64_linux.c3:11]: Yeah, I know. I just didn't feel like replicating all the subdirectories etc multiple times. If we have too many of them we'll make separate directories.

File: std/aarch64_linux/syscalls.c3 (hunk 1 above)
Comment (antroseco) [reply std/aarch64_linux/syscalls.c3]: Done

File: std/arm_linux/runtime.S (hunk 2 above)
Comment (DaveDuck321) [reply std/arm_linux/runtime.S:53]: Ok, that makes sense, no need then. I thought the kernel might not bother saving them given that it already has a different calling convention.

File: std/sys/mman.c3 [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    addr: iptr,
+    length: isize,
+    prot: u32,
+    flags: u32,
+    fd: int,
+    offset: off_t
+) -> iptr = {
+    return sys_mmap(addr, length, as_arithmetic(prot), as_arithmetic(flags),
+                    fd, offset);
+}
```
Comment (DaveDuck321) [inline std/sys/mman.c3:31]: Is this purely so the interface of `sys_mmap` matches the Linux calling convention or does it also help us with platform compatibility?

**Review by DaveDuck321 — APPROVED** [review]

File: std/sys/mman.c3 [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+// TODO: support enums (or similar)
+
+function SYS_PROT_READ  : () -> u32 = { return 0x1; } // Page can be read.
+function SYS_PROT_WRITE : () -> u32 = { return 0x2; } // Page can be written.
+function SYS_PROT_EXEC  : () -> u32 = { return 0x4; } // Page can be executed.
+function SYS_PROT_NONE  : () -> u32 = { return 0x0; } // Page can not be accessed.
+
+function SYS_MAP_SHARED          : () -> u32 = { return 0x1; } // Share changes.
+function SYS_MAP_PRIVATE         : () -> u32 = { return 0x2; } // Changes are private.
+function SYS_MAP_SHARED_VALIDATE : () -> u32 = { return 0x3; } // Share changes and validate extension flags.
```
Comment (DaveDuck321) [inline std/sys/mman.c3:17]: This is going to conflict with a similar change I made in #101. I like this one more tho, lets keep this

File: std/arm_linux/runtime.S (hunk 2 above)
Comment (antroseco) [reply std/arm_linux/runtime.S:53]: To be honest, not really sure what the kernel is supposed to do. Linux claims that only `lr` is clobbererd (https://github.com/torvalds/linux/blob/93f5de5f648d2b1ce3540a4ac71756d4a852dc23/tools/include/nolibc/arch-arm.h#L57) and indeed, the syscall handler starts by pushing `r1-r12` to the stack. On the other hand, glibc saves `r4-r7` only (not `lr`) (https://github.com/bminor/glibc/blob/4290aed05135ae4c0272006442d147f2155e70d7/sysdeps/unix/sysv/linux/arm/syscall.S#L26). So... I think I won't save any other registers... as both implementations seem to work just fine... If this turns out to be a problem, then it's an easy fix.

**Pull request merged**